app/services/DataGenerator.scala

```diff
@@ -0,0 +1,250 @@
+/*
+ * Copyright 2021 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package services
+
+import java.util.concurrent.TimeUnit
+
+import config.ApplicationConfig
+import javax.inject.{Inject, Singleton}
+import metrics.Metrics
+import models.{ERSFileProcessingException, SchemeData, SchemeInfo}
+import play.api.Logger
+import play.api.mvc.Request
+import services.ERSTemplatesInfo.ersSheets
+import services.audit.AuditEvents
+import services.validation.ErsValidator
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.services.validation.DataValidator
+import uk.gov.hmrc.services.validation.models.ValidationError
+import utils.ErrorResponseMessages
+
+import scala.collection.mutable.ListBuffer
+import scala.concurrent.ExecutionContext
+import scala.util.{Failure, Success, Try}
+
+@Singleton
+class DataGenerator @Inject()(auditEvents: AuditEvents,
+                              config: ApplicationConfig)(
+  implicit val ec: ExecutionContext) extends DataParser with Metrics {
+
+  val defaultChunkSize: Int = 10000
+  private[services] val ersSheetsClone: Map[String, SheetInfo] = ersSheets
+
+  def getErrors(iterator: Iterator[String])(implicit schemeInfo: SchemeInfo, hc: HeaderCarrier, request: Request[_]): ListBuffer[SchemeData] = {
+    var rowNum = 0
+    implicit var sheetName: String = ""
+    var sheetColSize = 0
+    val schemeData: ListBuffer[SchemeData] = ListBuffer()
+    var validator: DataValidator = ERSValidationConfigs.defValidator
+
+    def incRowNum() = rowNum = rowNum + 1
+
+    val startTime = System.currentTimeMillis()
+
+    def checkForMissingHeaders(rowNum: Int) = {
+      if (rowNum > 0 && rowNum < 9) {
+        throw ERSFileProcessingException(
+          s"${ErrorResponseMessages.dataParserIncorrectHeader}",
+          s"${ErrorResponseMessages.dataParserIncorrectHeader}")
+      }
+    }
+
+    while (iterator.hasNext) {
+
+      val row = iterator.next()
+      //Logger.debug(" Data before  parsing ---> " + row)
```

Comment:
Commented code. Remove if not required.

Reply:
removed - we didn't modify anything in that method btw, it's only ods-related